### PR TITLE
MM-8477 Add interactive menus to message attachments

### PR DESCRIPTION
**Pull request by jwilander**
#### Summary
Add interactive menus to message attachments.

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-8477

#### Checklist
- [x] Ran `make check-style` to check for style errors (required for all pull requests)
- [x] Ran `make test` to ensure unit and component tests passed
- [x] Added or updated unit tests (required for all new features)
- [x] Has server changes (https://github.com/mattermost/mattermost-server/pull/9285)
- [x] Has redux changes (https://github.com/mattermost/mattermost-redux/pull/610)
- [x] Has UI changes

### CODE REVIEW

**Comment by jwilander** [comment]
Waiting on the server PR to be merged and then I can set up the test server with an integration running to test with.

**Comment by jwilander** [comment]
@jasonblais I've added three posts on the test server in town square that you can use to test. I also set up an integration to respond to the menus in that post with ephemeral messages.

@asaadmahmood this is ready for styling, you'll probably want to do most of it in `action_menu.jsx`. You can pass a class to the div holding the textbox here: https://github.com/mattermost/mattermost-webapp/pull/1612/files#diff-5361cea2c8f7bd5c31344018e173eec6R71

**Comment by jasonblais** [comment]
@asaadmahmood Early next week is fine. Merge deadline for this PR is next Tuesday.

**Comment by jasonblais** [comment]
We can test the functionality in the meantime.

**Comment by asaadmahmood** [comment]
Cool, will have a look at it on Sunday, or probably Monday then hopefully.

**Comment by jwilander** [comment]
FYI if the test server gets destroyed I'll need to re set up the integration before it'll work again

**Comment by jasonblais** [comment]
Sounds good, @asaadmahmood @lindy65 @lindalumitchell let's not remove/delete the test server until all testing is completed, so we don't have to ask Joram to set up the integration multiple times.

@asaadmahmood And Monday works fine 👍

**Comment by lindy65** [comment]
@jwilander @jasonblais 

Thanks, I've tested using sysadmin credentials.

1. Using the 3 sample posts Joram created in Town Square, each behaves as expected showing a drop-down with a scrollable (if options are long enough) selection.  

2. What I am not certain about is whether I should be able to go back and select another option?  If I am supposed to be able to, I noticed that:
 - only `action options` sample message gives me the multiple options to select from again
 - `action users` and `action channels` sample messages only show the option I first chose when I click on them a second time
(Note: If I refresh the page, my previous selections are cleared, including ephemeral messages and I am able to select fresh options from the multiple option scrollable lists)

3. I'm not sure what the expected should be in the ephemeral messages regarding `action users` and `action channels` displaying strings (e.g. `ID cx5u3og81bn58b74nok64keu7a!`) as opposed to the user or channel name?  `Action options` seems to display the option name and not a string.

image

4. What is the expected for viewing in "compact" mode?  I'm seeing the message attachment in full

image

5. Not sure whether it's expected to be able to search for the options you've selected in the dropdown menu?  (No results are returned for me)

6. Not sure whether there will be a scenario where people reply to these messages where there was not original message text - just the attachment?  If so, is it expected that there is no `commented on...` line above the reply?

image

**Comment by jasonblais** [comment]
Thanks @lindy65, really appreciate the help testing. 

----

@jwilander Please help
 - Add the `action options` dropdown back (that's missing in Town Square for me)
 - Review questions 2, 3 and 6 from Lindy. I have responses to them below as well.

----

Will also answer your questions Lindy:

1. 👍 
2. For `action users` and `action channels` I think that's expected, since that's what's typed inside the box. If I start clearing characters, more results are shown that match the search term.

Though not sure why that is inconsistent with `action options`. Will try it out once it's been added to the spinmint.

image

3. I assume these are controlled by the integration itself, but I'll let Joram confirm.
4. Yeah that's expected 👍 
5. This is expected. Given they're not posted as messages in this case. But the integration creator can return the option you selected as an actual message (not as an ephemeral message), which can then be searched.
6. I think you have covered a bug here indeed. Though it's related to the ephemeral message that appears above the reply. Steps:
 - Choose an option from a dropdown from the bot message
 - Observe the resulting ephemeral message from the integration
 - Reply to the bot message
 - Observe no "commented on..." line

image

(This is on the message that does contain text above the message attachment)

 - Remove the ephemeral message
 - Observe the "commented on..." line appear

image

I'm not sure if this is a bug on `master`, but I wasn't able to reproduce this on pre-release.

**Comment by jasonblais** [comment]
Some additional notes:

——

Asaad:

7. There is less space in the RHS, so we'll want to make sure the CSS/UI covers cases when the option is really long. It breaks in the attached example.

image

8. After selecting an option, I get confused the my cursor stays inside the textbox. I'm like "do I have to do something else here". This is especially true when the message is higher up in the channel and I don't see the ephemeral message..

Should we set focus back to the message box after an option is selected? That might solve potential confusion in this case. 

——

Joram:

9. Does the integration return anything if I type an "incorrect" option and hit ENTER? It would be nice if there was an ephemeral message along the lines of `[User Entered Value] isn't one of the available options. Please try again.` 

This might be up to the integration creator to do. Not sure.

image

10. For users, I tried with `@` and it didn't return any results. Same with `~` for channels. I realize it's the same with the channel switcher for users, so this question might be out of scope for this PR (or a note to Asaad).

11. Can the options you choose stay after a refresh? I feel they should, but not sure if that’s technically possible. I suspect the message dropdowns don’t store anything when you select an option but instead just sends a request.

—— 

12. Bug ticket not caused by this PR: https://mattermost.atlassian.net/browse/MM-11848

**Comment by jwilander** [comment]
2. The options show up again if you click on the textbox somewhere near the start because the autocomplete only considers what is behind the caret for the suggestion list. For the options case, they all start with `Option*`, so that's why they show up there but not for users if you start the click earlier. This is just how our suggestion box works everywhere in the product.

3. That's up to the integration.

6. Yeah that's a bug, but not one introduced by this feature.

9. No, I followed how the buttons work in that case and they do not return anything for failed requests.

10. Yeah it's autocompleting without the `~` and `@` everywhere except for the mention autocompletes. We can probably make it work with both but I think that's beyond scope for this PR.

11. Correct, there's no easy way to record the option so it stays past refresh or across multiple devices. Slack behaves the same.

**Comment by jasonblais** [comment]
2 - Makes sense, let's keep the current behaviour as-is

3 - +1

6 - @lindy65 can you help see if you can get repro steps on CIs or pre-release, then file a separate ticket?

9 - `[User Entered Value] isn't one of the available options. Please try again.` works as the text @jwilander. Will it be shown each time a request fails?

10 - OK, can be in V2 if we hear people asking about it.

11 - Yeah I saw that too, I suspect people will start asking for it but that can be in V2.

**Comment by jwilander** [comment]
9 - Whoops sorry, I thought you meant if there was a failed request because of the integration not being available or it returned an error. The behaviour is right now we don't send any requests for "incorrect" options and there's no feedback the option is incorrect. That's the same as Slack. I can still add the ephemeral message (or it can just be in-line text appearing saying that's not an option) but it'll be a bit more work.

**Comment by jasonblais** [comment]
9 - Yeah I realize it's the same as Slack but they show "No items matched [User Entered Value]" in the autocomplete in that case, so it's more obvious. Is that something we can do too, or is adding an in-line text easier? 

image

**Comment by jasonblais** [comment]
^why I suggested the in-line or ephemeral text is because my guess is adding the "No items matched [User Entered Value]" text won't be easy given the autocomplete component doesn't do that right now

**Comment by jwilander** [comment]
9 - I can probably add the same "No items match" entry to the suggestion box, but it'll be at least 2 mana

**Comment by jasonblais** [comment]
9 - Okay, thanks. Can we do that, or at least create a separate ticket to do that after this PR is merged for tomorrow's feature cut.

**Comment by jwilander** [comment]
9 - I can create a ticket for it

**Comment by jwilander** [comment]
9 - https://mattermost.atlassian.net/browse/MM-11850

**Comment by asaadmahmood** [comment]
8. We can decide whether to focus it to the message area or not, though we should in any case blur (remove focus) from the input box.

@jasonblais I have another concern, there is no option for the user to change his selected option correctly. If we clicks on the input again, only the options that match the selected option, or only the selected option are shown.

If a user taps on the input again, we should remove the selected value from the input box, and show all the options, though we should still have the previously selected option highlighted in the suggestions list similar to Slack.

And if the user does not select any new option and clicks outside to close the suggestion list, then his previously selected option should return.

**Comment by jasonblais** [comment]
@asaadmahmood Thank you, will talk with Joram about above.

From UX perspective, what is your recommendation on 8? Remove focus from the input box?

**Comment by asaadmahmood** [comment]
@jasonblais Yes, wondering if we should also have a green checkmark to indicate that the action was successful. Though I think since we add a system message, that probably is sufficient.

**Comment by lindy65** [comment]
@jasonblais 

6) Reproduced on pre-release. Ticket: https://mattermost.atlassian.net/browse/MM-11854

**Comment by jwilander** [comment]
@asaadmahmood after your changes to `at_mention_provider.jsx` I'm seeing it pull the desctipions/nicknames to the right when I type `@` in the post textbox

screen shot 2018-08-27 at 15 15 26

**Comment by jwilander** [comment]
Ok, pushed my changes. Want me to re-create the test server now or wait until Asaad's fix for the above issue is in? @jasonblais

**Comment by jasonblais** [comment]
@jwilander When should we get this to dev review in order to reasonably merge by end of Tuesday?

I'd prefer to test it all at once after @asaadmahmood's fix, since your two fixes were minor. The UI changes are more significant.

**Comment by jwilander** [comment]
Soon, as if there is developer feedback I may need to make some changes. We should probably assign dev reviews now to get the ball rolling.

**Comment by jwilander** [comment]
I should note that Asaad's UI changes are done for the most part and would be available for testing. The one issue I mentioned doesn't affect the new feature

**Comment by jasonblais** [comment]
@jwilander Let's generate a new spinmint in that case and get testing underway.

**Comment by jwilander** [comment]
@jasonblais @lindy65 test server up and running with the integration set back up.

Note there is an unrelated bug on master causing: `TypeError: this.props.actions.getRedirectLocation is not a function` to appear in the console.

**Comment by jasonblais** [comment]
@asaadmahmood The channel autocomplete is also broken due to the changes (channel purpose appears next to the channel name instead of below it)

Observed:
image

Expected:
image

**Comment by asaadmahmood** [comment]
@jwilander @jasonblais Should be fixed.
@jasonblais I think we should still have some indication on the controller itself that you selected an option, like the checkmark or something.

`✓ Submitted` or `✓ Success`. Or something.

Imagine if this control is way above in the center channel and the user is scrolled up while using it. We do not take him to the bottom to show the system message, and that is fine, but we should show some indication on the component itself because the feedback at the bottom is not visible for the user at that point.

**Comment by lindy65** [comment]
@asaadmahmood - not sure whether this is an issue?

image

**Comment by asaadmahmood** [comment]
@lindy65 Definitely an issue, submitting a fix.

**Comment by lindy65** [comment]
@jwilander @jasonblais 

Not sure whether this is an issue or expected?  The top item in the dropdown menu (in all action boxes) remains highlighted even if I mouse over a selection lower down.  I am also not able to use up / down arrows to make a selection.  e.g.

image

**Comment by jasonblais** [comment]
@lindy65 That's how autocomplete works today (see attached for user autocomplete)

image

However, I think there is indeed room to improve. Can you help create a ticket for it and queue it for the UX team to discuss?

**Comment by jasonblais** [comment]
@asaadmahmood 

> I think we should still have some indication on the controller itself that you selected an option, like the checkmark or something.

Agree that's something to consider.
1) Can you help with a mockup of what this would look like
2) Are you okay shipping this feature without it, or do you think it's a blocker?

**Comment by asaadmahmood** [comment]
@jasonblais something like this.
screen shot 2018-08-28 at 5 06 15 pm

**Comment by jwilander** [comment]
Server ready to be tested

**Comment by jasonblais** [comment]
Noticed a minor issue: No space between username and display name following latest changes

image

**Comment by lindy65** [comment]
Ticket queued for UX team to discuss autocomplete: https://mattermost.atlassian.net/browse/MM-11888

**Comment by jasonblais** [comment]
@asaadmahmood If you have the time, hoping to get some of your help with this PR this week, specifically
 - https://github.com/mattermost/mattermost-webapp/pull/1612#issuecomment-418869334
 - https://mattermost.atlassian.net/browse/MM-11933

Hoping we can get this to dev review soon.

**Comment by jasonblais** [comment]
^ @asaadmahmood :)

**Comment by asaadmahmood** [comment]
@jasonblais Done.

**Comment by jwilander** [comment]
I'll rebase and update the test server

**Comment by jasonblais** [comment]
@asaadmahmood The feedback `Submitted` is broken on RHS. Propose having it below the dropdown similar to mobile view

image

Else LGTM

**Comment by lindy65** [comment]
Thanks @jwilander @jasonblais @asaadmahmood 

I've retested all previous issues found as well as everything else I could think of and have a few questions / comments below:

1. When clicking reply on one of the posts, after I have made a selection in the center panel, should the selection be visible in the RHS?  At the moment, I'm not seeing any of the selections visible in the RHS.  (Also, as JB mentioned, the "Submitted" feedback is broken in the RHS.)

image

2. Related to 1) - When clicking "reply" - if I _shouldn't_ see the selection in the RHS (that I chose in the center panel), if I make another selection in the RHS, should it update in the center panel?  

image

3. Selections with long names are cut off in the RHS

image

image

4. Channel descriptions are cut off in the center panel

image

**Comment by jasonblais** [comment]
Good questions Lindy!

Will defer to Joram on what is possible with 1 & 2.

3 & 4 are correct, but then noticed the same issue on regular autocompletes. Asaad may have thoughts on it though.

image

**Comment by jasonblais** [comment]
Thanks @asaadmahmood 👍 

@lindy65 New spinmint up with the integrations ready. I've tested, +1 from my end. Let me know if you want to do any additional testing. Note that 3/4 from your notes above are addressed separately as it affects autocompletes in general.

**Comment by lindy65** [comment]
Thanks @jwilander @jasonblais @asaadmahmood 

1. Joram, a question on for future QA "expected" behavior:
> Regarding 1/2, I didn't write any code that would mirror the selection from the center to the RHS but I suppose we need that to work otherwise it will be confusing to the user. I'll add support for that
>This will also result in the selection persisting across channel switch
- Selections will now persist across channel switch in center panel and RHS but will not after refresh, correct?
- Are these selections "user specific" or does everyone in the channel see them?

2. Noted that selection changes made in the RHS are now updated in the center panel 👍 

3. Ephemeral messages are missing on all posts - did something change?

4. Just a question about the "Submitted" feedback button...  If a user makes a change to their selection, a new ephemeral message posts (not at the moment because they're broken?) BUT the "Submitted" button remains static...  I was wondering whether the button should do something to confirm re-submission?  Like fade out and re-appear - not sure what's possible... (On the desktop app, when you are saving settings, "Saving..." appears while the new setting is being saved but disappears once done.  I was thinking of something like this with the button remaining after update).

**Comment by jwilander** [comment]
1. Yes, persist across channel switch but not refresh. They are user specific.
3. Yes, Jason re-created the test server so the integration wasn't running. Nothing to worry about

**Comment by jasonblais** [comment]
Created a separate ticket for 4 after discussing with Asaad: https://mattermost.atlassian.net/browse/MM-12235

This has now passed UX, PM and QA review, queuing for dev review next.

File: components/post_view/message_attachments/action_menu/action_menu.jsx [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+            value = selected.id;
+        } else {
+            displayText = selected.text;
+            value = selected.value;
+        }
+
+        this.props.actions.selectAttachmentMenuAction(this.props.postId, this.props.action.id, this.props.action.data_source, displayText, value);
+
+        if (this.suggestionRef) {
+            requestAnimationFrame(() => this.suggestionRef.blur());
```
[the remark this answers is not in the document]
Comment (jwilander) [reply components/post_view/message_attachments/action_menu/action_menu.jsx]: Good call, will update

**Comment by jwilander** [comment]
@hmhealey can you re-review?

**Comment by hmhealey** [comment]
Sorry, not sure how I missed re-reviewing this one. Looks good, but it needs a rebase

**Pull request merged**